File: leetcode/0023-merge-k-sorted-lists.cc
// https://leetcode.com/problems/merge-k-sorted-lists/
// Hard

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
#include<vector>
#include<queue>
using namespace std;

struct myComp { 
    constexpr bool operator()( 
        pair<int, int> const& a, 
        pair<int, int> const& b) 
        const noexcept 
    { 
        return a.first > b.first; 
    } 
};

class Solution {

public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode * sortedList = nullptr;
        ListNode * head = nullptr;
        priority_queue<pair<int, int>, vector<pair<int, int>>, myComp> pq;
        int size = lists.size();
        for (int i = 0; i < size; ++i) {
            if(lists[i]) pq.push({lists[i]->val, i});
        }
        pair<int, int> curr;
        while (!pq.empty()) {
            curr = pq.top();
            pq.pop();
            if (sortedList) {
                sortedList->next = lists[curr.second];
                sortedList = sortedList->next;
            } else {
                sortedList = lists[curr.second];
                head = sortedList;
            }
            if (lists[curr.second] && lists[curr.second]->next) {
                lists[curr.second] = lists[curr.second]->next;
                pq.push({lists[curr.second]->val, curr.second});
            }
        }
        return head;
    }
};

File: leetcode/python/0023-merge-k-sorted-lists.py
# https://leetcode.com/problems/merge-k-sorted-lists/
# Medium

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        h = []
        for i, l in enumerate(lists):
            if l:
                h.append((l.val, i))
        ret = head = ListNode()
        heapify(h)
        while len(h) > 0:
            smallest = lists[h[0][1]]
            head.next = smallest
            head = head.next
            if smallest.next != None:
                lists[h[0][1]] = smallest.next
                heapreplace(h, (smallest.next.val, h[0][1]))
            else:
                heappop(h)
        return ret.next